### PR TITLE
Implemented homework 4

### PR DESCRIPTION
**Pull request by stekkko**
Usefull links below:
-- Homework 4 http://www.seas.upenn.edu/~cis194/spring13/hw/04-higher-order.pdf
-- Homeworks http://www.seas.upenn.edu/~cis194/spring13/lectures.html

P.S. I`m not sure my insert function works well

### CODE REVIEW

File: src/HW4.hs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+module HW4 where 
+
+--exercise 1
+fun1 :: [Integer] -> Integer
+fun1 = product . map (subtract 2) . filter even
```
Comment (chshersh) [inline src/HW4.hs:5]: Nice implementation!

File: src/HW4.hs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+checkN :: Integer -> Integer
+checkN n = if even n then n `div` 2 else n * 3 + 1
+
+fun2 :: Integer -> Integer
+fun2 = sum . filter even . takeWhile (> 1) . iterate checkN
```
Comment (chshersh) [inline src/HW4.hs]: `fun2` is better than `fun2'`. It's a good approach to give names to long functions. What can be improved: you could put `checkN` function under `where` section for `fun2` function to not spoil global namespace.

File: src/HW4.hs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  | otherwise = let x@(Node d1 _ _ _) = insert el r
+                in Node (d1 + 1) l val x
+                      
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insert Leaf
+
+--exercise 3
+xor :: [Bool] -> Bool
+xor = odd . length . filter id
```
Comment (chshersh) [inline src/HW4.hs]: `xor` of two values is just `/=` operator. So you can implement this function simpler, using this fact :smirk:

File: src/HW4.hs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr ((:) . f) []
+
+--exercise 4
+seiveSundaram :: Int -> [Int]
+seiveSundaram n = (map ((+1) . (*2)) . filter (`notElem` badList n)) [1..n]
```
Comment (chshersh) [inline src/HW4.hs]: This is nice implementation!

But you can implement your function simpler using the fact that Haskell is a lazy language:

* http://slides.com/fp-ctd/lecture-2#/49

**Review by chshersh — COMMENTED** [review]
This is very nice! Looks like you're become pretty fluent with `(.)` operator and other concepts of Haskell! :slightly_smiling_face:

File: src/HW4.hs (hunk 2 above)
Comment (stekkko) [reply src/HW4.hs]: fun2 :: Integer -> Integer
    fun2 = sum . filter even . takeWhile (> 1) . iterate checkN
      where checkN n = if even n then n `div` 2 else n * 3 + 1

Is it ok?

Comment (chshersh) [reply src/HW4.hs]: @MINER9000 Yeah, that's okay

File: src/HW4.hs (hunk 4 above)
Comment (stekkko) [reply src/HW4.hs]: Well,  I can't do it simpler right now. I just leave it like that .

**Review by chshersh — APPROVED** [review]
Nice!

**Pull request merged**